PyTAPS/listener.py

```diff
@@ -0,0 +1,208 @@
+import asyncio
+import ssl
+from .connection import Connection
+from .securityParameters import SecurityParameters
+from .transportProperties import *
+from .endpoint import LocalEndpoint, RemoteEndpoint
+from .utility import *
+from .transports import *
+
+color = "cyan"
+
+class Listener():
+    """The TAPS listener class.
+
+    Attributes:
+        preconnection (Preconnection, required):
+                Preconnection object from which this Connection
+                object was created.
+    """
+    def __init__(self, preconnection):
+                # Initializations
+                self.local_endpoint = preconnection.local_endpoint
+                self.remote_endpoint = preconnection.remote_endpoint
+                self.transport_properties = preconnection.transport_properties
+                self.security_parameters = preconnection.security_parameters
+                self.loop = preconnection.loop
+                self.active = preconnection.active
+                self.framer = preconnection.framer
+                self.security_context = None
+                self.set_callbacks(preconnection)
+
+    async def start_listener(self):
+        """ method wrapped by listen
+        """
+        print_time("Starting listener.", color)
+
+        # Create set of candidate protocols
+        candidate_set = self.create_candidates()
+
+        # If the candidate set is empty issue an InitiateError cb
+        if not candidate_set:
+            print_time("Protocol selection Error occured.", color)
+            if self.initiate_error:
+                self.loop.create_task(self.initiate_error())
+            return
+
+        # If security_parameters were given, initialize ssl context
+        if self.security_parameters:
+            self.security_context = ssl.create_default_context(
+                                                ssl.Purpose.CLIENT_AUTH)
+            if self.security_parameters.identity:
+                print_time("Identity: " +
+                           str(self.security_parameters.identity))
+                self.security_context.load_cert_chain(
+                                        self.security_parameters.identity)
+            for cert in self.security_parameters.trustedCA:
+                self.security_context.load_verify_locations(cert)
+        # Attempt to set up the appropriate listener for the candidate protocol
```

Comment:
Is the implementation still able to listen using both TCP and UDP? Or does this require two separate Listener objects?
Maybe this is worth mentioning somewhere as a comment and in the docs.

Reply:
Yes, at the moment the listener would only work for either protocol. Wether we want it to be this way is something to discuss I guess.

Reply:
I checked the Arch and API drafts and I don't see anything that prevents a Listener from accepting both TCP and UDP to create ConnectionReceived events if the Transport Properties allow TCP and UDP. So I don't see why a restriction to either TCP or UDP would be necessary.
The discussion in this issue seems to point the same way: https://github.com/ietf-tapswg/api-drafts/issues/206

For this PR, I'd be fine with marking this as FIXME or something, but we should consider fixing it in the hackathon, where we're hopefully going to work on racing anyway. For example, this will reappear when we start to listen on different local interfaces with the same Listener.

Reply:
It is now back to how it was (one listener object for multiple different underlying protocols)